### PR TITLE
Add GGUF files validation and files endpoint

### PR DESCRIPTION
**Pull request by harsh97**

### CODE REVIEW

**Review by VipulMascarenhas — CHANGES_REQUESTED** [review]
few changes requested:
- can you change the target branch to feature/aquav1.0.3 instead of main?
- can we rename `arch` to `platform` instead?
- some tests are failing, might be good to fix them before we merge these changes.

**Review by VipulMascarenhas — COMMENTED** [review]
left a few comments.

File: ads/aqua/common/utils.py [hunk 1]
```diff
@@ -228,6 +230,29 @@ def load_config(file_path: str, config_file_name: str, **kwargs) -> dict:
...
+    Returns:
+    - A list of file paths matching the specified extension.
+    """
+    # Ensure the extension is prefixed with a dot if not already
+
+    signer = default_signer()
+
+    if not extension.startswith('.'):
+        extension = '.' + extension
+    fs = ocifs.OCIFileSystem(**signer)
```
Comment (VipulMascarenhas) [inline ads/aqua/common/utils.py:250]: we can reuse `list_objects` from `object_storage_details.py` instead.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -45,7 +45,7 @@ repos:
       rev: v8.18.4
       hooks:
           - id: gitleaks
-            exclude: .github/workflows/reusable-actions/set-dummy-conf.yml
+            exclude: .github/workflows/reusable-actions/set-dummy-conf.yml|./tests/operators/common/test_load_data.py
```
[the remark this answers is not in the document]
Comment (harsh97) [reply .pre-commit-config.yaml:48]: line =       self.data_spec.vault_secret_id = 'mock_secret_id' was being flagged as secret violation

File: ads/aqua/model/model.py [hunk 3]
```diff
@@ -218,7 +219,7 @@ def get(self, model_id: str, load_model_card: Optional[bool] = True) -> "AquaMod
         ).value
 
         aqua_model_attributes = dict(
-            **self._process_model(ds_model, self.region),
+            **self.to_aqua_model(ds_model, self.region),
```
[the remark this answers is not in the document]
Comment (harsh97) [reply ads/aqua/model/model.py]: have created a new method instead

File: ads/aqua/model/model.py [hunk 4]
```diff
@@ -425,6 +426,15 @@ def _process_model(
             == READY_TO_IMPORT_STATUS
         )
 
+        gpu_supported = (
+            freeform_tags.get(Tags.GPU_MODEL, "").upper == GPU_SUPPORTED_STATUS
```
[the remark this answers is not in the document]
Comment (harsh97) [reply ads/aqua/model/model.py]: as discussed platform will now be derrived from continer config and model_format

File: ads/aqua/modeldeployment/deployment.py [hunk 5]
```diff
@@ -523,10 +526,21 @@ def get_deployment_config(self, model_id: str) -> Dict:
...
+            full_config = load_config(
                 AQUA_CONFIG_FOLDER,
                 config_file_name=AQUA_MODEL_DEPLOYMENT_CONFIG_DEFAULTS,
             )
+            model: AquaModelSummary = AquaModelSummary(
+                **AquaModelApp.to_aqua_model(
+                    self.get_source(model_id), region=self.region
+                )
+            )
+            # Todo: Revisit this logic when we have multi arch support
```
[the remark this answers is not in the document]
Comment (harsh97) [reply ads/aqua/modeldeployment/deployment.py]: discussed offline and removed this change

File: ads/aqua/common/utils.py (hunk 1 above)
Comment (harsh97) [reply ads/aqua/common/utils.py]: addressed. Thanks!

File: ads/aqua/config/deployment_config_defaults.json [hunk 6]
```diff
@@ -1,9 +1,34 @@
 {
-  "shape": [
-    "VM.GPU.A10.1",
-    "VM.GPU.A10.2",
-    "BM.GPU.A10.4",
-    "BM.GPU4.8",
-    "BM.GPU.A100-v2.8"
-  ]
+  "CPU": {
```
[the remark this answers is not in the document]
Comment (harsh97) [reply ads/aqua/config/deployment_config_defaults.json]: updated

File: ads/aqua/common/utils.py [hunk 7]
```diff
@@ -228,6 +231,34 @@ def load_config(file_path: str, config_file_name: str, **kwargs) -> dict:
...
+    Parameters:
+    - oss_path: The path to the directory where files are located.
+    - extension: The file extension to filter by (e.g., 'txt' for text files).
+
+    Returns:
+    - A list of file paths matching the specified extension.
+    """
+
+    oss_client = ObjectStorageDetails.from_path(oss_path)
+    signer = default_signer()
```
Comment (VipulMascarenhas) [inline ads/aqua/common/utils.py:247]: nit: remove redundant code

File: ads/aqua/model/model.py [hunk 8]
```diff
@@ -53,6 +55,13 @@
 from ads.model import DataScienceModel
 from ads.model.model_metadata import ModelCustomMetadata, ModelCustomMetadataItem
 from ads.telemetry import telemetry
+from oci.data_science.models import JobRun, Model
+
+
+class ModelValidationResult:
```
Comment (VipulMascarenhas) [inline ads/aqua/model/model.py:61]: nit: this can go inside `ads.aqua.model.entities`

File: ads/aqua/model/model.py [hunk 9]
```diff
@@ -423,6 +444,24 @@ def _process_model(
             == READY_TO_IMPORT_STATUS
         )
 
+        inference_containers = AquaContainerConfig.from_container_index_json().inference
```
Comment (VipulMascarenhas) [inline ads/aqua/model/model.py:447]: we should check if this adds any increase in latency for get() API as any file read operation will introduce significant delay.

File: ads/aqua/model/model.py [hunk 10]
```diff
@@ -676,6 +726,150 @@ def _create_model_catalog_entry(
         logger.debug(model)
         return model
 
+    @staticmethod
+    def get_model_files(os_path: str, model_format: ModelFormat) -> [str]:
```
Comment (VipulMascarenhas) [inline ads/aqua/model/model.py:730]: nit: add docstrings

File: ads/aqua/model/model.py [hunk 11]
```diff
@@ -676,6 +726,150 @@ def _create_model_catalog_entry(
...
+            else:
+                model_files.append(AQUA_MODEL_ARTIFACT_CONFIG)
+
+        if model_format == ModelFormat.GGUF:
+            model_files.extend(
+                list_os_files_with_extension(oss_path=os_path, extension=".gguf")
+            )
+        return model_files
+
+    def validate_model_config(
```
Comment (VipulMascarenhas) [inline ads/aqua/model/model.py:749]: nit: add docstrings

**Review by VipulMascarenhas — APPROVED** [review]
changes look good, add minor comments. I'm curious to see if the API latency increases for some of the calls we're making to fetch info from config files.

File: ads/aqua/model/model.py (hunk 9 above)
Comment (harsh97) [reply ads/aqua/model/model.py:447]: I have added a cache on the file read in utils function!!

File: ads/aqua/model/model.py (hunk 11 above)
Comment (harsh97) [reply ads/aqua/model/model.py:749]: done

File: ads/aqua/model/model.py (hunk 10 above)
Comment (harsh97) [reply ads/aqua/model/model.py:730]: done

File: ads/aqua/model/model.py (hunk 8 above)
Comment (harsh97) [reply ads/aqua/model/model.py:61]: done

File: ads/aqua/common/utils.py (hunk 7 above)
Comment (harsh97) [reply ads/aqua/common/utils.py:247]: done

**Pull request merged**